### PR TITLE
Agent simple inverse queries

### LINKED ISSUE #298: Implement simple inverse queries of `provider` and `receiver` on agent records
A [PR to VF GraphQL](https://lab.allmende.io/valueflows/vf-schemas/vf-graphql/-/issues/100) has to be made first before any of the simple [inverse query edges](https://www.valueflo.ws/specification/inverses/) can be implemented.

You can ignore
- any fields involving `X.inScopeOf` for now, pending #297 
- any fields involving `EconomicResource.primaryAccountable` for now, pending #133 

This only includes "inverse queries", not "other queries".

The focus here is:
`Commitment.provider` and `Commitment.receiver`
`Intent.provider` and `Intent.receiver`
`EconomicEvent.provider` and `EconomicEvent.receiver`

- [ ] comment out [the lines so far in `Agent`](https://github.com/h-REA/hREA/blob/5f01158a78266d4c0e96b8ff34a3a0047f0a7b1f/modules/vf-graphql-holochain/resolvers/agent.ts#L46-L85) which will be uncommented later in #297 and #133
- [ ] add `Agent.commitmentsAsProvider`
- [ ] add `Agent.commitmentsAsReceiver`
- [ ] add `Agent.intentsAsProvider`
- [ ] add `Agent.intentsAsReceiver`
- [ ] add `Agent.economicEventsAsProvider`
- [ ] add `Agent.economicEventsAsReceiver`

### For `Commitment.provider`:
We will need these to be implemented:
https://github.com/h-REA/hREA/blob/5f01158a78266d4c0e96b8ff34a3a0047f0a7b1f/zomes/rea_commitment/zome_idx_planning/src/lib.rs#L35-L36

### For `Intent.provider`:
we already have
https://github.com/h-REA/hREA/blob/5f01158a78266d4c0e96b8ff34a3a0047f0a7b1f/zomes/rea_intent/zome_idx_planning/src/lib.rs#L30-L32
and
https://github.com/h-REA/hREA/blob/7e3d5ad7a0849f8c4996d9c767ce64c486058870/zomes/rea_intent/lib/src/lib.rs#L37-L42

is it then just a matter of hooking up the resolver function? Is this relationship tested?

### PR DESCRIPTION
**Pull request by weswalla**
closes #298

### CODE REVIEW

**Comment by weswalla** [comment]
@Connoropolous @pospi there seems to be a gql level error when trying to query the `commitmentsAsProvider`/`commitmentsAsReceiver` fields of `agent, as done here:
https://github.com/h-REA/hREA/blob/7f651d3424651fa16dbf38862a1276e0610b5699/test/agent/test_agent_links.js#L87
and here:
https://github.com/h-REA/hREA/blob/7f651d3424651fa16dbf38862a1276e0610b5699/test/agent/test_agent_links.js#L96

The resolver functions aren't even executing:
https://github.com/h-REA/hREA/blob/7f651d3424651fa16dbf38862a1276e0610b5699/modules/vf-graphql-holochain/resolvers/agent.ts#L57

which makes me suspect it has something to do with gql syntax and the fact that in the schema we define inputs: https://lab.allmende.io/valueflows/vf-schemas/vf-graphql/-/blob/sprout/lib/schemas/bridging/agent.commitment.gql#L53-54

Any ideas, do I need to define more input parameters to:
https://github.com/h-REA/hREA/blob/7f651d3424651fa16dbf38862a1276e0610b5699/modules/vf-graphql-holochain/resolvers/agent.ts#L57

**Comment by weswalla** [comment]
I confirmed that the indexing is working properly and calling `query_commitments` returns the relevant `CommitmentConnection` object. So this is defintely a gql related issue, maybe something to do with the fact that Agent is an interface then `Person` and `Organization` implement that interface?

EDIT: actually the error does seem to be coming from zome side, something to do with `construct_response`

**Comment by pospi** [comment]
Glad you figured it out!

A couple of issues with those parameters- I would parameterize `commitmentsAsReceiver` too.

The other issue is passing all of them- it's actually considered undefined behavior to send all of them. You can omit `after` and `before` since we are starting from the end, and should only pass `last` to get the latest, not `before`. See https://relay.dev/graphql/connections.htm#sec-Pagination-algorithm for more info.

**Comment by weswalla** [comment]
@pospi there seems to be an error with `get_link_fields`:
https://github.com/h-REA/hREA/blob/fc7d80565356e95e010610d5b866aa28e4166473/zomes/rea_agent/lib/src/lib.rs#L139-L165

It should be returning a vec of `CommitmentAddresses`, but it isn't. I'm assuming the error is either from `read_index!()` or `create_index!()` on the commitment side of things: https://github.com/h-REA/hREA/blob/fc7d80565356e95e010610d5b866aa28e4166473/zomes/rea_commitment/lib/src/lib.rs#L36

Although I'm a bit stumped as to why its not working.

To be more specific, I printed the output of the `person` query (in the api code), which should contain the `commitmentsAsProvider` and other related fields, but they are excluded from the object, which means these are empty vecs in `get_link_fields`

**Comment by weswalla** [comment]
@Connoropolous @pospi this is ready for review and should close #298

File: zomes/rea_commitment/lib/src/lib.rs [hunk 1]
```diff
@@ -164,6 +148,32 @@ pub fn handle_update_commitment<S>(entry_def_id: S, commitment: UpdateRequest) -
...
+                .provider(new_value.as_slice())
+                .not(prev_value.as_slice()),
+            agent.commitments_as_provider(&base_address)
+        );
+        hdk::prelude::debug!("handle_update_commitment::provider index {:?}", e);
+    }
+    if new_entry.receiver != prev_entry.receiver {
+        // let new_value = match &new_entry.receiver { Some(val) => vec![val.to_owned()], None => vec![] };
+        let new_value = vec![new_entry.receiver.to_owned()];
+        // let prev_value = match &prev_entry.receiver { Some(val) => vec![val.to_owned()], None => vec![] };
```
Comment (Connoropolous) [inline zomes/rea_commitment/lib/src/lib.rs]:
```suggestion
```

File: zomes/rea_commitment/lib/src/lib.rs [hunk 2]
```diff
@@ -164,6 +148,32 @@ pub fn handle_update_commitment<S>(entry_def_id: S, commitment: UpdateRequest) -
     //     );
     //     hdk::prelude::debug!("handle_update_commitment::in_scope_of index {:?}", e);
     // }
+    if new_entry.provider != prev_entry.provider {
+        // let new_value = match &new_entry.provider { Some(val) => vec![val.to_owned()], None => vec![] };
+        let new_value = vec![new_entry.provider.to_owned()];
+        // let prev_value = match &prev_entry.provider { Some(val) => vec![val.to_owned()], None => vec![] };
```
Comment (Connoropolous) [inline zomes/rea_commitment/lib/src/lib.rs]:
```suggestion
```

File: zomes/rea_commitment/lib/src/lib.rs [hunk 3]
```diff
@@ -164,6 +148,32 @@ pub fn handle_update_commitment<S>(entry_def_id: S, commitment: UpdateRequest) -
     //     );
     //     hdk::prelude::debug!("handle_update_commitment::in_scope_of index {:?}", e);
     // }
+    if new_entry.provider != prev_entry.provider {
+        // let new_value = match &new_entry.provider { Some(val) => vec![val.to_owned()], None => vec![] };
```
Comment (Connoropolous) [inline zomes/rea_commitment/lib/src/lib.rs]:
```suggestion
```

File: zomes/rea_agent/rpc/src/lib.rs [hunk 4]
```diff
@@ -114,11 +132,17 @@ impl<'a> UpdateRequest {
...
+    // pub inventoried_economic_resources: Option<EconomicResourceAddress>,
+    // pub plans: Option<PlanAddress>,
+    // pub processes: Option<ProcessAddress>,
+    // pub proposals: Option<ProposalAddress>,
+    pub commitments_as_provider: Option<CommitmentAddress>,
+    pub commitments_as_receiver: Option<CommitmentAddress>,
+    pub intents_as_provider: Option<IntentAddress>,
+    pub intents_as_receiver: Option<IntentAddress>,
+    pub economic_events_as_provider: Option<IntentAddress>,
+    pub economic_events_as_receiver: Option<IntentAddress>,
```
Comment (Connoropolous) [inline zomes/rea_agent/rpc/src/lib.rs]:
```suggestion
    pub economic_events_as_provider: Option<EconomicEventAddress>,
    pub economic_events_as_receiver: Option<EconomicEventAddress>,
```

File: test/agent/test_agent_links.js [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+// const testIntentProps = {
+//   action: 'raise',
+//   resourceClassifiedAs: ['some-resource-type'],
+//   resourceQuantity: { hasNumericalValue: 1, hasUnit: mockIdentifier() },
+// }
+// const testEconomicEventProps = {
+//   action: 'raise',
+//   resourceClassifiedAs: ['some-resource-type'],
+//   resourceQuantity: { hasNumericalValue: 1, hasUnit: mockIdentifier() },
+// }
```
Comment (Connoropolous) [inline test/agent/test_agent_links.js]:
```suggestion
```

File: zomes/rea_commitment/lib/src/lib.rs [hunk 6]
```diff
@@ -164,6 +148,32 @@ pub fn handle_update_commitment<S>(entry_def_id: S, commitment: UpdateRequest) -
...
+        let e = update_index!(
+            commitment
+                .provider(new_value.as_slice())
+                .not(prev_value.as_slice()),
+            agent.commitments_as_provider(&base_address)
+        );
+        hdk::prelude::debug!("handle_update_commitment::provider index {:?}", e);
+    }
+    if new_entry.receiver != prev_entry.receiver {
+        // let new_value = match &new_entry.receiver { Some(val) => vec![val.to_owned()], None => vec![] };
```
Comment (Connoropolous) [inline zomes/rea_commitment/lib/src/lib.rs]:
```suggestion
```

**Review by Connoropolous — APPROVED** [review]
updating the vf-graphql broke one or two things, which I identified by running the full test suite, but I fixed those. Other than that, it's just removing the commented out code, and one bug where `IntentAddress` should have been `EconomicEventAddress`

File: test/plan/test_plan_links.js [hunk 7]
```diff
@@ -72,6 +72,8 @@ test('Plan links & queries', async (t) => {
   const pId = resp.data.process.process.id
   const cId = resp.data.commitment.commitment.id
 
+  console.log('argh', JSON.stringify(resp))
```
Comment (pospi) [inline test/plan/test_plan_links.js:75]: just this one to clean up, then I think we're good to merge

**Pull request merged**